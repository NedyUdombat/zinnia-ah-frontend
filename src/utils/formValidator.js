/**
 * validate inputs
 * usage: const errors = validate({ email: 'user@mail.com', password: 'aNewSecret' }, exitsingErrors);
 */
export const validate = (values, validationErrors) => {
  const newErrors = [];
  const oldErrors = [...validationErrors];
  const keys = Object.keys(values);

  const username = value => {
    if (isFieldEmpty('username', value)) {
      return;
    }

    if (trim(value).length < 3) {
      newErrors.push(lengthError('username', 3));
    } else {
      const index = oldErrors.indexOf(lengthError('username', 3));

      oldErrors.includes(lengthError('username', 3))
        ? oldErrors.splice(index, 1)
        : '';
    }

    if (!trim(value).match(/^([a-zA-Z0-9]+)$/)) {
      newErrors.push(alphanumError('username'));
    } else {
      const index = oldErrors.indexOf(alphanumError('username'));

      oldErrors.includes(alphanumError('username'))
        ? oldErrors.splice(index, 1)
        : '';
    }
  };

  const password = value => {
    if (isFieldEmpty('password', value)) {
      return;
    }

    if (trim(value).length < 8) {
      newErrors.push(lengthError('password', 8));
    } else {
      const index = oldErrors.indexOf(lengthError('password', 8));

      oldErrors.includes(lengthError('password', 8))
        ? oldErrors.splice(index, 1)
        : '';
    }

    if (!trim(value).match(/^([a-zA-Z0-9]+)$/)) {
      newErrors.push(alphanumError('password'));
    } else {
      const index = oldErrors.indexOf(alphanumError('password'));

      oldErrors.includes(alphanumError('password'))
        ? oldErrors.splice(index, 1)
        : '';
    }
  };

  const email = value => {
    if (isFieldEmpty('email', value)) {
      return;
    }

    const pattern = /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$/;
    const isEmail = trim(value).match(pattern);
    if (!isEmail) {
      newErrors.push(emailError);
    } else {
      const index = oldErrors.indexOf(emailError);

      oldErrors.includes(emailError) ? oldErrors.splice(index, 1) : '';
    }
  };

  const isFieldEmpty = (field, value) => {
    if (trim(value).length < 1) {
      newErrors.push(isRequiredError(field));
      return true;
    }
    const index = oldErrors.indexOf(isRequiredError(field));

    oldErrors.includes(isRequiredError(field))
      ? oldErrors.splice(index, 1)
      : '';

    return false;
  };

  keys.forEach(key => {
    switch (key) {
      case 'username':
        username(values[key]);
        break;

      case 'email':
        email(values[key]);
        break;

      case 'password':
        password(values[key]);
        break;

      default:
        break;
    }
  });

  return [...new Set(oldErrors.concat(newErrors))];
};

export const emailError = 'email must be a valid email';

export const alphanumError = field => {
  return `${field} can only contain alphanumerics`;
};

export const isRequiredError = field => {
  return `${field} is required`;
};

export const lengthError = (field, requiredLength) => {
  return `${field} must be at least ${requiredLength} characters long`;
};

export const trim = word => word.replace(/([ ])+/g, '');
