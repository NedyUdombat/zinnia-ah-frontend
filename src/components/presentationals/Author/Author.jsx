import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

//utils
import { DEFAULT_USER_IMAGE_URL } from '../../../utils/config';

//components
import Avatar from '../Avatar/Avatar';
import Button from '../Button/Button';

//stylesheets
import '../ProfileSidebar/ProfileSidebar.scss';
const Author = ({ profile }) => {
  return profile ? (
    <div className="container" data-test="author">
      <div className="avatar-div">
        <Avatar url={DEFAULT_USER_IMAGE_URL} className="avatar-xl" />
      </div>
      <p className="text-center fullname">{`${profile.firstName} ${
        profile.lastName
      }`}</p>
      <Link to={`@${profile.username}`}>
        <p className="text-center username">@{profile.username}</p>
      </Link>
      <Button className="btn-white" value="FOLLOW" type="submit" />
    </div>
  ) : null;
};

Author.propTypes = {
  profile: PropTypes.shape({
    firstName: PropTypes.string.isRequired,
    lastName: PropTypes.string.isRequired,
    username: PropTypes.string.isRequired,
  }).isRequired,
};

export default Author;
