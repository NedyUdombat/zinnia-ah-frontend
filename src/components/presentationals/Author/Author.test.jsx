import React from 'react';
import { shallow } from 'enzyme';
import checkPropTypes from 'check-prop-types';

import Author from './Author';
import { findByTestAttribute } from '../../../utils/testHelpers';

const props = {
  profile: {
    lastName: '',
    firstName: '',
    username: '',
  },
};

describe('Author', () => {
  describe('Render', () => {
    it('should render without errors', () => {
      const component = shallow(<Author {...props} />);
      expect(component).toMatchSnapshot();
      const wrapper = findByTestAttribute(component, 'author');
      expect(wrapper.length).toBe(1);
    });
  });

  describe('Proptypes', () => {
    it('should receive props', () => {
      const expectedProps = {
        ...props,
      };

      const propsErrors = checkPropTypes(
        Author.propTypes,
        expectedProps,
        'props',
        Author,
      );

      expect(propsErrors).toBeUndefined();
    });
  });
});
