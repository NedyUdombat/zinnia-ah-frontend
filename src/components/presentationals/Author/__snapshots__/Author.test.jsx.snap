// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Author Render should render without errors 1`] = `
<div
  className="container"
  data-test="author"
>
  <div
    className="avatar-div"
  >
    <Avatar
      className="avatar-xl"
      url="https://res.cloudinary.com/nedy123/image/upload/v1557149386/avatar_pupqrh.svg"
    />
  </div>
  <p
    className="text-center fullname"
  >
     
  </p>
  <Link
    to="@"
  >
    <p
      className="text-center username"
    >
      @
    </p>
  </Link>
  <Button
    className="btn-white"
    type="submit"
    value="FOLLOW"
  />
</div>
`;
